subiquity/ui/views/ubuntu_advantage.py

```diff
@@ -89,11 +140,107 @@ def on_cancel(_: UbuntuAdvantageForm):
+        if token:
+            checking_token_overlay = CheckingUAToken(self)
+            self.show_overlay(checking_token_overlay,
+                              width=checking_token_overlay.width,
+                              min_width=None)
+            self.controller.check_token(token)
+        else:
+            self.controller.done(token)
 
     def cancel(self) -> None:
         """ Called when the user presses the Back button. """
         self.controller.cancel()
+
+    def show_invalid_token(self) -> None:
+        """ Display an overlay that indicates that the user-supplied token is
+        invalid. """
+        self.remove_overlay()
+        self.show_stretchy_overlay(
+                SomethingFailed(self,
+                                "Invalid token.",
+                                "The Ubuntu Advantage token that you provided"
+                                " is invalid. Please make sure that you typed"
+                                " your token correctly."))
+
+    def show_expired_token(self) -> None:
+        """ Display an overlay that indicates that the user-supplied token has
+        expired. """
+        self.remove_overlay()
+        self.show_stretchy_overlay(
+                SomethingFailed(self,
+                                "Token expired.",
+                                "The Ubuntu Advantage token that you provided"
+                                " has expired. Please use a different token."))
+
+    def show_unknown_error(self) -> None:
+        """ Display an overlay that indicates that we were unable to retrieve
+        the subscription information. Reasons can be multiple include lack of
+        network connection, temporary service unavailability, API issue ...
+        The user is prompted to continue anyway or go back.
+        """
+        self.remove_overlay()
+        self.show_stretchy_overlay(ContinueAnywayWidget(self))
+
+    def show_available_services(self, services: dict) -> None:
+        """ Display an overlay with the list of services that will be enabled
+        via Ubuntu Advantage subscription. After the user confirms, the next we
+        will quit the current view and move on. """
+        self.remove_overlay()
+        self.show_stretchy_overlay(ShowServicesWidget(self, services))
+
+
+class ShowServicesWidget(Stretchy):
+    """ Widget to show the available services for UA subscription. """
+    def __init__(self, parent: UbuntuAdvantageView, services: list):
+        """ Initializes the widget by including the list of services as a
+        bullet-point list. """
+        self.parent = parent
+
+        ok = ok_btn(label=_("OK"), on_press=self.ok)
+
+        title = _("Available Services")
+        header = _("List of services that will be enabled through your "
+                   "Ubuntu Advantage subscription:")
+
+        widgets: List[Widget] = [Text(header)]
+        widgets.extend([Text(f"* {svc['description']}") for svc in services])
```

Comment:
I think this dialog should have a blank line between the header and services and between the services and the button. And the button should be focused by default.

Actually, it could probably also have some text along the lines of "you can enable these services using the 'ua' command line tool" between the list and the button.

Reply:
I think the first two comments apply to the other dialogs too.

Reply:
Thanks!
I created https://github.com/canonical/subiquity/pull/1157 to address this comment.